### PR TITLE
Fix Stalker Arm64

### PR DESCRIPTION
**Pull request by akiannillo**
* implemented load real register into for X29/X30 registers
* added push|pull q registers in stalker code when calling functions
* new performance test for stalker arm 64

### CODE REVIEW

File: tests/core/interceptor.c [hunk 1]
```diff
@@ -120,6 +119,19 @@ INTERCEPTOR_TESTCASE (attach_to_special_function)
...
+  gint (* pthread_key_create_impl) (pthread_key_t *key, void (*destructor)(void*));
+  pthread_key_t key;
+
+  pthread_key_create_impl = GSIZE_TO_POINTER(
+      gum_module_find_export_by_name ("libc.so", "pthread_key_create"));
+
+  interceptor_fixture_attach_listener (fixture, 0, pthread_key_create_impl, '>', '<');
+
+  g_assert_cmpint (pthread_key_create_impl (&key, NULL), ==, 0);
+}
```
Comment (oleavr) [inline tests/core/interceptor.c]: Changes to this file are unrelated and should go in a separate PR.

File: gum/arch-arm64/gumarm64relocator.c [hunk 2]
```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+      }
+    }
+    else
+    {
+      n = 4;
+    }
 
+    cs_free (insn, 1);
     cs_close (&capstone);
   }
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Changes to this file are unrelated and should go in a separate PR.

File: gum/backend-arm64/gumstalker-arm64.c [hunk 3]
```diff
@@ -1332,6 +1331,12 @@ gum_exec_ctx_load_real_register_into (GumExecCtx * ctx,
   gint slot_in_the_stack;
   gint pos_in_the_slot;
 
+#if STALKER_DEBUG_LEVEL == 10
+  g_print("gum_exec_ctx_load_real_register_into - start\n");
+  g_print("\nsource: %d\n", source_register);
+  g_print("\ntarget: %d\n", target_register);
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: Missing space between function name and opening parenthesis on these three lines.

Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: Is this debug actually useful now that the problem has been fixed?

File: tests/core/arch-arm64/stalker-arm64.c [hunk 4]
```diff
@@ -188,6 +191,8 @@ STALKER_TESTCASE (unconditional_branch_reg)
 
   address = GUM_ADDRESS (gum_arm64_writer_cur (&cw));
   gum_arm64_writer_put_add_reg_reg_imm (&cw, ARM64_REG_X0, ARM64_REG_X0, 10);
+  if (reg == ARM64_REG_X29 || reg == ARM64_REG_X30)
+    gum_arm64_writer_put_pop_reg_reg(&cw, reg, ARM64_REG_XZR);
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]: Missing space between function name and opening parenthesis.

File: tests/core/arch-arm64/stalker-arm64.c [hunk 5]
```diff
@@ -439,6 +446,11 @@ STALKER_TESTCASE (follow_return)
       GUM_ARG_ADDRESS, fixture->stalker,
       GUM_ARG_ADDRESS, fixture->sink);
   gum_arm64_writer_put_pop_all_x_registers (&cw);
+  /*
+   * alternative for instruction RET X15
+   * gum_arm64_writer_put_mov_reg_reg(&cw, ARM64_REG_X15, ARM64_REG_X30);
+   * gum_arm64_writer_put_instruction(&cw, 0xD65F01E0);
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]: Missing space between function name and opening parenthesis on these two lines.

File: gum/backend-arm64/gumstalker-arm64.c [hunk 6]
```diff
@@ -1874,11 +1918,16 @@ gum_exec_block_write_ret_transfer_code (GumExecBlock * block,
   gum_exec_block_close_prolog (block, gc);
   gum_exec_block_open_prolog (block, GUM_PROLOG_MINIMAL, gc);
 
+  gum_exec_ctx_load_real_register_into (block->ctx, ARM64_REG_X16, ret_reg, 0,
+      gc);
+
   gum_arm64_writer_put_push_all_x_registers (cw);
+  gum_arm64_writer_put_push_all_q_registers (cw);
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: We don't save registers here on x86, because that's what the opened prolog has already done. Is this a bug?

File: gum/backend-arm64/gumstalker-arm64.c [hunk 7]
```diff
@@ -2060,6 +2109,7 @@ gum_exec_block_write_event_submit_code (GumExecBlock * block,
   beach_label = cw->code + 1;
 
   gum_arm64_writer_put_push_all_x_registers (cw);
+  gum_arm64_writer_put_push_all_q_registers (cw);
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: We don't save registers here on x86, because that's what the opened prolog has already done. Is this a bug?

File: gum/backend-arm64/gumstalker-arm64.c [hunk 8]
```diff
@@ -2079,10 +2130,12 @@ gum_exec_block_write_event_submit_code (GumExecBlock * block,
     gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X14, beach_label);
 
     gum_arm64_writer_put_push_all_x_registers (cw);
+    gum_arm64_writer_put_push_all_q_registers (cw);
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: We don't save registers here on x86, because that's what the opened prolog has already done. Is this a bug?

File: gum/backend-arm64/gumstalker-arm64.c (hunk 3 above)
Comment (akiannillo) [inline gum/backend-arm64/gumstalker-arm64.c]: no, I can remove them.

File: gum/backend-arm64/gumstalker-arm64.c (hunk 6 above)
Comment (akiannillo) [inline gum/backend-arm64/gumstalker-arm64.c]: if I remove them, the performance test get stuck again.

File: gum/backend-arm64/gumstalker-arm64.c (hunk 7 above)
Comment (akiannillo) [inline gum/backend-arm64/gumstalker-arm64.c]: idem here, test crashes without this.

File: gum/backend-arm64/gumstalker-arm64.c (hunk 8 above)
Comment (akiannillo) [inline gum/backend-arm64/gumstalker-arm64.c]: here it actually works if I remove them

File: gum/backend-arm64/gumstalker-arm64.c [hunk 9]
```diff
@@ -2068,7 +2094,8 @@ gum_exec_block_write_event_submit_code (GumExecBlock * block,
       GUM_ADDRESS (block->ctx->sink_process_impl), 2,
       GUM_ARG_ADDRESS, block->ctx->sink,
       GUM_ARG_REGISTER, ARM64_REG_X15);
-  gum_arm64_writer_put_pop_all_x_registers (cw);
+  gum_arm64_writer_put_pop_reg_reg (cw, STALKER_REG_CTX, ARM64_REG_X15);
+
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: Two blank lines here.
